File: src/api/schemas/scheduler.py
"""调度系统 API Schema。"""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field

ScheduleKind = Literal["daily_at", "weekdays_at", "cron"]
ScheduleHint = Literal["morning", "pre_open", "post_close", "anytime"]


class ScheduleCommandItem(BaseModel):
    command_key: str
    label: str
    typer_group: str
    typer_command: str
    category: str
    schedule_hint: ScheduleHint
    run_on_trading_day: bool
    referenced_by: list[str] = Field(default_factory=list)
    is_referenced: bool = False


class ScheduleJobItem(BaseModel):
    job_key: str
    name: str
    schedule_kind: ScheduleKind
    schedule_time: str
    cron_expr: Optional[str] = None
    run_on_trading_day: bool = False
    enabled: bool = True
    command_keys: list[str] = Field(default_factory=list)
    command_count: int = 0
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class ScheduleJobCreateRequest(BaseModel):
    job_key: str = Field(..., min_length=1, max_length=64)
    name: str = Field(..., min_length=1, max_length=128)
    schedule_kind: ScheduleKind = "daily_at"
    schedule_time: str = Field(..., pattern=r"^\d{2}:\d{2}$")
    cron_expr: Optional[str] = None
    run_on_trading_day: bool = False
    enabled: bool = True
    command_keys: list[str] = Field(..., min_length=1)


class ScheduleJobUpdateRequest(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=128)
    schedule_kind: Optional[ScheduleKind] = None
    schedule_time: Optional[str] = Field(default=None, pattern=r"^\d{2}:\d{2}$")
    cron_expr: Optional[str] = None
    run_on_trading_day: Optional[bool] = None
    enabled: Optional[bool] = None
    command_keys: Optional[list[str]] = None


class ScheduleRunStepItem(BaseModel):
    step_id: int
    command_key: str
    sort_order: int
    status: str
    saved_count: Optional[int] = None
    message: Optional[str] = None
    started_at: Optional[str] = None
    finished_at: Optional[str] = None


class ScheduleRunItem(BaseModel):
    run_id: int
    job_id: Optional[int] = None
    job_key: Optional[str] = None
    triggered_by: str
    status: str
    started_at: Optional[str] = None
    finished_at: Optional[str] = None
    error_message: Optional[str] = None
    steps: list[ScheduleRunStepItem] = Field(default_factory=list)


class ScheduleRunListRequest(BaseModel):
    job_key: Optional[str] = None
    page: int = Field(default=1, ge=1)
    count: int = Field(default=20, ge=1, le=500)


class ScheduleRunListResponse(BaseModel):
    items: list[ScheduleRunItem]
    total: int


class ScheduleOverviewResponse(BaseModel):
    command_total: int
    command_referenced_count: int
    command_unreferenced_count: int
    commands: list[ScheduleCommandItem]
    jobs_enabled_count: int
    last_run_at: Optional[str] = None
    recent_runs: list[ScheduleRunItem] = Field(default_factory=list)


class ScheduleRunTriggerResponse(BaseModel):
    run_id: int

File: src/web/admin/src/types/scheduler.ts
export type ScheduleHint = 'morning' | 'pre_open' | 'post_close' | 'anytime';
export type ScheduleKind = 'daily_at' | 'weekdays_at' | 'cron';

export interface ScheduleCommandItem {
  command_key: string;
  label: string;
  typer_group: string;
  typer_command: string;
  category: string;
  schedule_hint: ScheduleHint;
  run_on_trading_day: boolean;
  referenced_by: string[];
  is_referenced: boolean;
}

export interface ScheduleJobItem {
  job_key: string;
  name: string;
  schedule_kind: ScheduleKind;
  schedule_time: string;
  cron_expr?: string | null;
  run_on_trading_day: boolean;
  enabled: boolean;
  command_keys: string[];
  command_count: number;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface ScheduleRunStepItem {
  step_id: number;
  command_key: string;
  sort_order: number;
  status: string;
  saved_count?: number | null;
  message?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
}

export interface ScheduleRunItem {
  run_id: number;
  job_id?: number | null;
  job_key?: string | null;
  triggered_by: string;
  status: string;
  started_at?: string | null;
  finished_at?: string | null;
  error_message?: string | null;
  steps?: ScheduleRunStepItem[];
}

export interface ScheduleOverviewResponse {
  command_total: number;
  command_referenced_count: number;
  command_unreferenced_count: number;
  commands: ScheduleCommandItem[];
  jobs_enabled_count: number;
  last_run_at?: string | null;
  recent_runs: ScheduleRunItem[];
}

export interface ScheduleJobCreateRequest {
  job_key: string;
  name: string;
  schedule_kind: ScheduleKind;
  schedule_time: string;
  cron_expr?: string | null;
  run_on_trading_day?: boolean;
  enabled?: boolean;
  command_keys: string[];
}

export interface ScheduleJobUpdateRequest {
  name?: string;
  schedule_kind?: ScheduleKind;
  schedule_time?: string;
  cron_expr?: string | null;
  run_on_trading_day?: boolean;
  enabled?: boolean;
  command_keys?: string[];
}

export interface ScheduleRunListResponse {
  items: ScheduleRunItem[];
  total: number;
}

export interface ScheduleRunTriggerResponse {
  run_id: number;
}

export const SCHEDULE_PRESETS: { label: string; kind: ScheduleKind; time: string }[] = [
  { label: '每天早上 06:00', kind: 'daily_at', time: '06:00' },
  { label: '开盘前 09:25', kind: 'weekdays_at', time: '09:25' },
  { label: '收盘后 16:30', kind: 'weekdays_at', time: '16:30' },
  { label: '夜间低峰 02:00', kind: 'daily_at', time: '02:00' },
];

export const scheduleHintLabels: Record<ScheduleHint, string> = {
  morning: '早晨',
  pre_open: '开盘前',
  post_close: '收盘后',
  anytime: '任意时间',
};

export const scheduleKindLabels: Record<ScheduleKind, string> = {
  daily_at: '每天',
  weekdays_at: '工作日',
  cron: 'Cron',
};

export const scheduleRunStatusLabels: Record<string, string> = {
  running: '运行中',
  success: '成功',
  failed: '失败',
  partial: '部分成功',
  skipped: '已跳过',
};

export function formatScheduleRunStatus(status: string): string {
  return scheduleRunStatusLabels[status] ?? status;
}
